pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+
+
+
+Motivation (Tobias' version)
+============================
+
+**This section should explain why we think pattern matching is a good
+addition for Python.**
+
+Since Python is a dynamically typed language, Python code frequently
+has to deal with data that comes in different forms and shapes.  This,
+in turn, gives rise to a high degree of versatility by favouring a
+'duck typing' style, imposing only the bare minimum of requirements
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
```

Comment:
Careful, `match` can't differentiate between `List[int]` and `List[str]` either.

Reply:
Yes, it can due to the possibility of nested patterns.  It cannot do `List[str]` or `List[int]` directly, but you can do `tuple(int(), int(), int())`, say.  This is something that is clearly not possible with any of the other approaches.  But I might have to formulate it better.

In short: pattern matching is _deep matching_ as opposed to the _shallow matching_ of other mtehods.

Reply:
Sure, but the example doesn't quite live up to this claim without further explanation.

Reply:
Yes.  The text is unclear in its current form.